src/compression.rs

```diff
@@ -1189,13 +1180,63 @@ impl CompressionStepResult {
     }
 }
 
+/// the complete result of a single step of compression, this is a somewhat expensive data structure
+/// to create.
+#[derive(Debug, Clone)]
+pub struct CompressionStepResult {
+    pub inv: Invention,
+    pub done: FinishedPattern,
+    pub dc_inv_str: String,
+    pub train_data: CompressionStepData,
+    pub test_data: Option<CompressionStepData>,
+}
+
+impl CompressionStepResult {
+    fn new(done: FinishedPattern, train_programs_node: Id, test_programs_node: Option<Id>, inv_name: &str, shared: &mut SharedData, past_invs: &Vec<CompressionStepResult>, first_train_cost: Option<i32>, first_test_cost: Option<i32>) -> Self {
+
+        let inv = done.to_invention(inv_name, shared);
+
+        let train_data = CompressionStepData::new(done.clone(), train_programs_node, inv_name, shared, past_invs, first_train_cost);
+        let test_data = test_programs_node.map(|n| CompressionStepData::new(done.clone(), n, inv_name, shared, past_invs, first_test_cost));
+
+        let dc_inv_str: String = dc_inv_str(&inv, past_invs);
+
+        CompressionStepResult { inv, done, dc_inv_str, train_data, test_data }
+    }
+
+    pub fn json(&self) -> serde_json::Value {        
+
+        json!({            
+            "body": self.inv.body.to_string(),
+            "dreamcoder": self.dc_inv_str,
+            "arity": self.inv.arity,
+            "name": self.inv.name,
+            "utility": self.done.utility,
+            "train_result": self.train_data.json(&self.inv.name),
+            "test_result": if let Some(d) = &self.test_data { d.json(&self.inv.name).to_string() } else { "N/A".to_string() },
```

Comment:
See comment for line 96 in `compress.rs`.

Reply:
I see, this might be a place where cloning would be needed.

Oh I just found this that might help: [as_deref](https://doc.rust-lang.org/std/option/enum.Option.html#method.as_deref)

That converts from `&Option<T>` to `Option<&T>` so you should be able to do something like `self.test_data.as_deref().map()` to achieve the thing like `inspect` that you wanted?

Reply:
See top-level comment below.